proof/meta.go

```diff
@@ -85,6 +131,188 @@ func (m *MetaReveal) Validate() error {
+
+	// These bytes must match our metadata size constraints.
+	err := IsValidMetaSize(jBytes, MetaDataMaxSizeBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal checks internally if the JSON is valid.
+	err = json.Unmarshal(jBytes, &jMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	return jMeta, nil
+}
+
+// EncodeMetaJSON encodes a JSON object as bytes and checks that the resulting
+// bytes are below the maximum metadata size.
+func EncodeMetaJSON(jMeta map[string]interface{}) ([]byte, error) {
+	if jMeta == nil {
+		return []byte{}, nil
+	}
+
+	jBytes, err := json.Marshal(jMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	err = IsValidMetaSize(jBytes, MetaDataMaxSizeBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return jBytes, nil
+}
+
+// GetDecDisplay attempts to decode metadata as JSON and return the decimal
+// display value. If the metadata is not JSON, or the decimal display key is
+// missing, the default decimal display value of 0 is returned.
+func (m *MetaReveal) GetDecDisplay() (map[string]interface{}, uint32, error) {
+	// An empty meta reveal has an implicit decimal display value of 0.
+	if m == nil {
+		return nil, 0, nil
+	}
+
+	if m.Type != MetaJson {
+		return nil, 0, ErrNotJSON
+	}
+
+	// Data must first be valid metadata JSON.
+	metaJSON, err := DecodeMetaJSON(m.Data)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// Our default decimal display value is 0.
+	decDisplay, ok := metaJSON[MetadataDecDisplayKey]
+	if !ok {
+		return metaJSON, 0, ErrDecDisplayMissing
+	}
+
+	// We have to typecheck the value stored for the decimal display key;
+	// it could be another JSON type.
+	switch decDisplayVal := decDisplay.(type) {
+	case float64:
+		// Any Number-like values are stored in a float64. We need to
+		// range check the decoded value.
+		if decDisplayVal < 0 {
+			return nil, 0, fmt.Errorf("decoded decimal display "+
+				"value is negative: %v", decDisplayVal)
+		}
+
+		if decDisplayVal > float64(MaxDecDisplay) {
+			return nil, 0, fmt.Errorf("%w: %v",
+				ErrDecDisplayTooLarge, decDisplayVal)
+		}
+
+		return metaJSON, uint32(decDisplayVal), nil
+
+	default:
+		return nil, 0, ErrDecDisplayInvalidType
+	}
+}
+
+// SetDecDisplay attempts to set the decimal display value in existing JSON
+// metadata. It checks that the new metadata is below the maximum metadata size.
+func (m *MetaReveal) SetDecDisplay(decDisplay uint32) (*MetaReveal, error) {
+	err := IsValidDecDisplay(decDisplay)
+	if err != nil {
+		return nil, err
+	}
+
+	// Fetch the current decimal display value.
+	currentMetaJSON, currentDecDisplay, err := m.GetDecDisplay()
+	switch {
+	// The current metadata is valid JSON. Either no decimal display value
+	// is present, or it is but doesn't match the desired value.
+	case err == ErrDecDisplayMissing, currentDecDisplay != decDisplay:
+		// If the requested decimal display value is 0, we don't need to
+		// add the field at all, as that is the default.
+		if decDisplay == 0 {
```

Comment:
I think it's better if we always set it, even if it is the default `0` value. Otherwise API users have to also assume the default value which may be unclear compared to the field always being present